lib/connection.js

```diff
@@ -1,12 +1,126 @@
 'use strict';
 
 const common = require('metarhia-common');
-const EventEmitter = require('events');
+const { EventEmitter } = require('events');
 
-const Connection = function() {
+const parser = require('./parser.js');
+
+const Connection = function(transport) {
   EventEmitter.call(this);
+
+  this._transport = transport;
+  this._parcels = [];
+  this._chunks = [];
+
+  // Mechanism for collecting full data,
+  // cut by lower-level protocols
+  this._bytesToRead = 36;
+  this._data = Buffer.alloc(0);
+
+  this._transport.on('data', (data) => {
+    if (this._bytesToRead === 0) {
+      const structType = data.readUInt8();
+      this._bytesToRead = structType === 0 ? 16 : 12;
+    }
+    this._onRawData(data);
+  });
+
+  // Process handshake as a first structure
+  this.once('structure', this._onHandshake);
 };
 
 common.inherits(Connection, EventEmitter);
 
+// Mechanism for collecting data,
+// cut by lower-level protocols
+Connection.prototype._onRawData = function(data) {
+  if (data.length < this._bytesToRead) {
+    this._bytesToRead -= data.length;
+    this._data = Buffer.concat([this._data, data]);
+  } else {
+    const newDataChunk = data.slice(0, this._bytesToRead);
+    this._data = Buffer.concat([this._data, newDataChunk]);
+    data = data.slice(this._bytesToRead);
+
+    // Chunk structure header
+    if (this._data.readUInt8() === 1 && this._data.length === 12) {
+      const length = this._data.readUInt16LE(10); // payload length
```

Comment:
All binary manipulation should be placed in `parser.js` I think

Reply:
So we need some `readPayloadLength` function in our parser.